Enhance googlecloudmonitoringreceiver: Add Unit Tests for Metric Conversion and Improve Edge Case Coverage

scaletech-milan opened the pull request:
**Description:** 
Added comprehensive test cases for the entire Google Cloud Monitoring Receiver component.

**Link to tracking Issue:** 
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36815

**Testing:** 
Implemented test cases to validate:

- GCP Monitoring client connection.
- OpenTelemetry's built-in functions.
- Scraping logic.
- Metric conversion functions.

**Documentation:**
N/A

scaletech-milan committed:
enhancement:
- Added unit tests for validating metric conversion logic in googlecloudmonitoringreceiver
- Improved test coverage for edge cases in the receiver component

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver_test.go:
```diff
@@ -0,0 +1,872 @@
+	ctx := context.Background()
+
+	// Set up fake credentials for testing
+	os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", "./../googlecloudspannerreceiver/testdata/serviceAccount.json")
```
lets copy the serviceAccount into this component (in a testdata directory) to avoid cross-module interactions.

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver_test.go:
```diff
@@ -0,0 +1,872 @@
+
+	// Test the Start function
+	err := receiver.Start(ctx, nil)
+	errMsg := fmt.Errorf("%s", "failed to find default credentials: google: could not find default credentials. See https://cloud.google.com/docs/authentication/external/set-up-adc for more information")
```
lets avoid checking for specific error strings, as these may change over time.  Can we maybe just check for "credentials", or something less likely to change?

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver_test.go:
```diff
@@ -0,0 +1,872 @@
+
+	// Test the Start function
+	err = receiver.Start(ctx, nil)
+	errMsg := fmt.Errorf("%s", "failed to retrieve metric descriptors data: rpc error: code = NotFound desc = project not found")
```
Can we check the status code from the error using "google.golang.org/grpc/status" and "google.golang.org/grpc/codes"?

```golang
assert.Equal(t, status.Code(err), codes.NotFound)
```

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver_test.go:
```diff
@@ -0,0 +1,872 @@
+
+	// Test the Start function
+	err = receiver.Start(ctx, nil)
+	errMsg := fmt.Errorf("%s", "failed to retrieve metric descriptors data: rpc error: code = Unauthenticated desc = transport: per-RPC creds failed due to error: auth: cannot fetch token: 400")
```
same comment.  Lets check codes

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver_test.go:
```diff
@@ -0,0 +1,872 @@
+	metrics, err := receiver.Scrape(ctx)
+
+	// Assert
+	errMsg := fmt.Errorf("%s", "failed to retrieve time series data: rpc error: code = NotFound desc = project not found")
```
Same comment.  Check for code: NotFound.

dashpole reviewed (commented):
looks good overall.  Lets just make the tests less fragile.

scaletech-milan committed:
enhancement:
- Update testify error handling code
- Add sample serviceAccount.json inside testdata

scaletech-milan commented:
Updated all the changes

scaletech-milan committed:
Merge branch 'main' into googlecloudmonitoringreceiver-unit-tests

dashpole commented:
Looks good.  Just needs the CLA signed, and I think we are good to go

scaletech-milan commented:
> Looks good. Just needs the CLA signed, and I think we are good to go

i already proceed for CLA sign, waiting for approval.

scaletech-milan committed:
enhancement:
- Resolve make lint issues

scaletech-milan commented:
Can you please check now @dashpole

dashpole commented:
```
Run make checkapi
  make checkapi
  shell: /usr/bin/bash -e {0}
  env:
    TEST_RESULTS: testbed/tests/results/junit/results.xml
    SEGMENT_DOWNLOAD_TIMEOUT_MINS: [2](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/12414428726/job/34674820908?pr=36846#step:10:2)
go run cmd/checkapi/main.go .
Ignoring extension/observer per allowlist
Ignoring extension/opampcustommessages per allowlist
202[4](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/12414428726/job/34674820908?pr=36846#step:10:4)/12/19 19:28:44 receiver/googlecloudmonitoringreceiver has more than one function: "NewFactory,NewFakeMetricTestServer"
```

Consider making NewFakeMetricTestServer and other testing-related artifacts not exported (lowercase)

scaletech-milan committed:
enhancement:
- Run make checkapi

scaletech-milan committed:
Merge branch 'open-telemetry:main' into googlecloudmonitoringreceiver-unit-tests

scaletech-milan commented:
@TylerHelmuth @dashpole 
Can you please give workflow approval ?

scaletech-milan committed:
Merge branch 'main' into googlecloudmonitoringreceiver-unit-tests

dashpole commented:
--- a/receiver/googlecloudmonitoringreceiver/go.mod
+++ b/receiver/googlecloudmonitoringreceiver/go.mod
@@ -69,7 +69,7 @@ require (
 	golang.org/x/text v0.21.0 // indirect
 	google.golang.org/api v0.210.0
 	google.golang.org/genproto/googleapis/rpc v0.0.0-20241118233622-e639e219e697 // indirect
-	google.golang.org/grpc v1.69.0 // indirect
+	google.golang.org/grpc v1.69.0
 	google.golang.org/protobuf v1.36.0
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
go.mod/go.sum deps changes detected, please run "make gotidy" and commit the changes in this PR.

TylerHelmuth commented:
definitely needs an update from main
